### PR TITLE
 Reuse the original request, avoid the abnormal ContentLength

### PR DESCRIPTION
**Pull request by jim3ma**
The `http.NewRequest` did not set the right `ContentLength` due to the type of `inReq.Body`.
We need to change the input of `http.NewRequest` to `*bytes.Buffer`, `*bytes.Reader` or `*strings.Reader`.

BTW, we can just set `ContentLength` like this `outReq.ContentLength = inReq.ContentLength` to workaround.

### CODE REVIEW

**Comment by bemasc** [comment]
I think your "workaround" would be a better solution.  This approach seems like it would behave poorly for streaming POSTs (e.g. large file uploads), because it would require them to be totally buffered in RAM on the proxy.

**Comment by mholt** [comment]
Yeah I vote for the workaround; let the streaming stream, and avoid buffering.

**Comment by jim3ma** [comment]
Okay.
One question, why not reuse the original request ?

**Comment by sergeyfrolov** [comment]
@jim3ma I believe @bemasc and @mholt do suggest to reuse the original request and avoid buffering. I concur.

**Comment by sergeyfrolov** [comment]
@jim3ma many thanks for your contribution!

Have you tested this with various GET requests in the wild?

I also realized there is another thing we might want to take care of. Websites generally receive requests that only have path in RequestURI and look like this:
```
GET /path/url.jpg HTTP/1.1
Host: website.com
```
while proxy receives (and with this PR sends) requests with full absoluteURI:
```
GET http://website.com/path/url.jpg HTTP/1.1
Host: website.com
```
Requests with absoluteURI are actually [technically valid requests](https://www.w3.org/Protocols/rfc2616/rfc2616-sec5.html#sec5.1.2), to be sent directly to the `website.com`. Nevertheless, I think it might be a good idea to stick to overwhelmingly common form of GET requests with only path in it for 2 reasons:
1. Full URI gives out the fact that request was forwarded by a proxy.
2. A lot of things on the web don't go out of their way to completely conform to the standards on cases that are so uncommon.

**Comment by jim3ma** [comment]
@sergeyfrolov 
We need to do:
1. rebuild the URL from req.RequestURI and req.Host
2. set the original request's URL
3. reset the original request's RequestURI to ""

than the original request could be request as a client request.

File: forwardproxy.go [hunk 1]
```diff
@@ -284,7 +284,7 @@ func (fp *ForwardProxy) ServeHTTP(w http.ResponseWriter, r *http.Request) (int,
 			panic("There was a check for http version, yet it's incorrect")
 		}
 	} else {
-		outReq, err := fp.generateForwardRequest(r)
+		outReq, err := fp.decorateOriginalRequest(r)
```
Comment (bemasc) [inline forwardproxy.go]: Since this is explicitly a mutator, I think we should probably get rid of `outReq` entirely.

**Comment by sergeyfrolov** [comment]
@jim3ma many thanks for you contribution!

**Pull request merged**